Use integer keys

batiati opened the pull request:
### This PR

Makes `Table.Key` always an integer type, removing the need for the `compare_keys` function.

**Rationale:** 

Removes unnecessary extra comparisons performed by `std.math.order` when we need to check only one condition, e.g. `a > b` instead of `std.math.order(a, b) == .gt`.

Future benefits may be explored when dealing with integer keys.

**Results:**

- Binary size decreased from 11757K to 10934K. Removing the inline function resulted in a **823K** smaller binary.

- Slightly better performance for both sequential and random IDs:

<table>
<th>
<td>Before</td><td>After</td>
</th>
<tr>
<td>Sequential IDs</td>
<td>
<pre>
Benchmarking...
info: Benchmark running against { 127.0.0.1:44605 }
info(message_bus): connected to replica 0
1226 batches in 59.76 s
load offered = 1000000 tx/s
load accepted = 167324 tx/s
batch latency p00 = 0 ms
batch latency p10 = 13 ms
batch latency p20 = 14 ms
batch latency p30 = 19 ms
batch latency p40 = 19 ms
batch latency p50 = 20 ms
batch latency p60 = 23 ms
batch latency p70 = 23 ms
batch latency p80 = 24 ms
batch latency p90 = 25 ms
batch latency p100 = 2494 ms
transfer latency p00 = 0 ms
transfer latency p10 = 3242 ms
transfer latency p20 = 6370 ms
transfer latency p30 = 10733 ms
transfer latency p40 = 15543 ms
transfer latency p50 = 20723 ms
transfer latency p60 = 26259 ms
transfer latency p70 = 31955 ms
transfer latency p80 = 38095 ms
transfer latency p90 = 44121 ms
transfer latency p100 = 49773 ms
</pre>
</td>
<td>
<pre>
Benchmarking...
info: Benchmark running against { 127.0.0.1:47107 }
info(message_bus): connected to replica 0
1226 batches in 58.06 s
load offered = 1000000 tx/s
load accepted = 172247 tx/s
batch latency p00 = 0 ms
batch latency p10 = 12 ms
batch latency p20 = 13 ms
batch latency p30 = 15 ms
batch latency p40 = 19 ms
batch latency p50 = 20 ms
batch latency p60 = 22 ms
batch latency p70 = 23 ms
batch latency p80 = 24 ms
batch latency p90 = 25 ms
batch latency p100 = 2684 ms
transfer latency p00 = 0 ms
transfer latency p10 = 2675 ms
transfer latency p20 = 5603 ms
transfer latency p30 = 9744 ms
transfer latency p40 = 14511 ms
transfer latency p50 = 19412 ms
transfer latency p60 = 25255 ms
transfer latency p70 = 30772 ms
transfer latency p80 = 36665 ms
transfer latency p90 = 42643 ms
transfer latency p100 = 48067 ms
</pre>
</td>
</tr>
<tr>
<td>Random IDs</td>
<td>
<pre>
Benchmarking...
info: Benchmark running against { 127.0.0.1:36915 }
info(message_bus): connected to replica 0
1226 batches in 68.28 s
load offered = 1000000 tx/s
load accepted = 146456 tx/s
batch latency p00 = 0 ms
batch latency p10 = 18 ms
batch latency p20 = 19 ms
batch latency p30 = 22 ms
batch latency p40 = 26 ms
batch latency p50 = 28 ms
batch latency p60 = 30 ms
batch latency p70 = 31 ms
batch latency p80 = 32 ms
batch latency p90 = 34 ms
batch latency p100 = 2824 ms
transfer latency p00 = 0 ms
transfer latency p10 = 3396 ms
transfer latency p20 = 6709 ms
transfer latency p30 = 12640 ms
transfer latency p40 = 18089 ms
transfer latency p50 = 24672 ms
transfer latency p60 = 30580 ms
transfer latency p70 = 37002 ms
transfer latency p80 = 44130 ms
transfer latency p90 = 51227 ms
transfer latency p100 = 58288 ms
</pre>
</td>
<td>
<pre>
Benchmarking...
info: Benchmark running against { 127.0.0.1:45447 }
info(message_bus): connected to replica 0
1226 batches in 66.87 s
load offered = 1000000 tx/s
load accepted = 149544 tx/s
batch latency p00 = 0 ms
batch latency p10 = 17 ms
batch latency p20 = 19 ms
batch latency p30 = 23 ms
batch latency p40 = 26 ms
batch latency p50 = 28 ms
batch latency p60 = 29 ms
batch latency p70 = 31 ms
batch latency p80 = 32 ms
batch latency p90 = 34 ms
batch latency p100 = 2506 ms
transfer latency p00 = 0 ms
transfer latency p10 = 3035 ms
transfer latency p20 = 6561 ms
transfer latency p30 = 11519 ms
transfer latency p40 = 16549 ms
transfer latency p50 = 23144 ms
transfer latency p60 = 29478 ms
transfer latency p70 = 35383 ms
transfer latency p80 = 42422 ms
transfer latency p90 = 49794 ms
transfer latency p100 = 56878 ms
</pre>
</td>
</tr>
</table>

batiati commented on src/lsm/binary_search.zig:
```diff
@@ -43,9 +42,9 @@ pub fn binary_search_values_upsert_index(
         if (constants.verify) {
             assert(offset == 0 or
-                compare_keys(key_from_value(&values[offset - 1]), key) != .gt);
+                !(key_from_value(&values[offset - 1]) > key));
```
It's an excellent opportunity to put https://github.com/tigerbeetle/tigerbeetle/pull/1179#discussion_r1322873550 in practice, any thoughts?

| before | after |
--------|--------
`compare_keys(a,b) == .eq` | `a == b`
`compare_keys(a,b) == .lt` | `a < b`
`compare_keys(a,b) == .gt` | `a > b`
`compare_keys(a,b) != .eq` | `a != b`
`compare_keys(a,b) != .lt` | `!(a < b)`
`compare_keys(a,b) != .gt` | `!(a > b)`

In some cases, especially in asserts, `!(a > b)` reads better than `a <= b` imo.

matklad replied:
Heh, using the rationale from that comment, I would say that

```zig
assert(
  key_from_value(&values[offset - 1]) <= key
)
```

is the most readable version:

* it states the invariant positively 
* it follows the natural number-line order of comparison, where the lesser number is on the left

That is, using the same antecedent, I seem to arrive at the opposite conclusion :rofl: 


Either version is much more readable than the one with `.gt` though!

matklad replied:
Also, spelled this way, it's clear that the assertion is actually wrong! We want 

```zig
key_from_value(&values[offset - 1]) < key
```

here, as per invariant from the docs: 

```zig
key_value_from(values[i-1]) < key or i == 0
```

incidentally, I was thinking about these invariants during my PTO:

https://matklad.github.io/2023/10/06/what-is-an-invariant.html

matklad replied:
Ahh, no, I think the docs are incorrect: now that we have lower/upper in config, the invariant actually depends on which bound we are looking for. So I think we need to

* adjust the docs
* use different asserts depenring on `comptime switch (config.mode)` (the current assert works for both cases, but it is not as tight as it could be)

matklad replied:
>In some cases, especially in asserts, !(a > b) reads better than a <= b imo.

Could you give an intuition why `!(a > b)` reads better for you? I just can't see it, to me it's obviously a worse readable version, so I am wondering if I am missing some different way to look at comparisons?

batiati replied:
> Could you give an intuition why !(a > b) reads better for you?

Not a general case, but for example `assert(!(key > last_key))`, my inner voice reads it as "it cannot be greater than the last key", which I think is easier to reason than `assert(key <= last_key)`.
IMO dealing with `>` and `<` is natural, but `>=` and `<=` are more subject to subtle "off-by-one" errors.

However, I'm not strong about it. The main motivation for converting `compare_keys(a,b) != .gt` to `!(a > b)` was to make it easier to mechanically refactor the code. I'd be happier if we could define a canonical way and document it in the TigerStyle.

batiati replied:
Unrelated but relevant:

What is the expected semantics of `maybe(<condition>)`?
Should we state something that is likely to always happen, or should we state the rule's exception?

For example, after `message_bus.unref()` we likely have freed the message, but maybe not, so the condition should be `maybe(messages.references > 0)` or `maybe(messages.references == 0)`?

I'm asking that because we could introduce a new function called `assert_not` for that kind of usage like `assert_not(key > last_key)`.

matklad replied:
>What is the expected semantics of maybe(<condition>)?

The intended semantics is that the condition can be either true or false, and, once we have branch coverage, we should have tests tripping both sides.

I don't think we assign any  probabilities to maybe, so both `maybe(cond)` and `maybe(!cond)` are equivalent. Though, I think we usually state the more surprising thing in the `cond`, as a cognitive help for the reader.

matklad commented on src/lsm/composite_key.zig:
```diff
@@ -27,44 +27,43 @@ pub fn CompositeKey(comptime Field: type) type {
+        pub inline fn key_from_value(value: *const Self) Key {
+            return @bitCast(Self{
                 .field = value.field,
                 .timestamp = @as(u63, @truncate(value.timestamp)),
```
```suggestion
                .timestamp = @as(u63, @truncate(value.timestamp)),
                .field = value.field,
```

matklad commented on src/lsm/composite_key.zig:
```diff
@@ -27,44 +27,43 @@ pub fn CompositeKey(comptime Field: type) type {
+            const value: Self = @bitCast(key);
+            return Self{
+                .field = value.field,
+                .timestamp = value.timestamp | tombstone_bit,
```
This makes me wonder if perhaps tombstone could be the least significant bit of the timestamp, rather than the MSB. That way, when we do binary search, we won't have to mask, and instead use the integer as is. This in turn would allow us to parametrize binary search not by Key/Value types, but rather by:

1) size of value
2) offset of the key in value
3) size of the key

The benefit here would be:

* smaller code size
* potentially faster runtime

For the on-disk format, LSB tombstone feels like a win to me. However, I don't like this as an over-the-wire format --- we never allow tombstones on the wire, and an unset LSB would look weird.

OTOH, its always the replica who assigns timestamps, and I think we could always assign even timestamps? Ie, we won't even document that we use LSB for tombstones, it's just so happens that all alive objects get assigned even timestamps.

matklad commented on src/lsm/binary_search.zig:
```diff
@@ -43,9 +42,9 @@ pub fn binary_search_values_upsert_index(
+                !(key_from_value(&values[offset - 1]) > key));
             assert(offset + length == values.len or
-                compare_keys(key_from_value(&values[offset + length]), key) != .lt);
+                !(key_from_value(&values[offset + length]) < key));
```
This one is correct, but I'd spell this as 

```
key <= key_from_value(&values[offset + length])
```

matklad commented on src/lsm/composite_key.zig:
```diff
@@ -27,44 +27,43 @@ pub fn CompositeKey(comptime Field: type) type {
-        /// The most significant bit must be unset as it is used to indicate a tombstone.
-        timestamp: u64,
         /// [0]u8 as zero-sized-type workaround for https://github.com/ziglang/zig/issues/16394.
         padding: @TypeOf(pad) = pad,
```
padding first feels weird, and I think it actually is wrong, as it breaks alignment. If Field is u128, then, with padding first, it is alligned only as good as u64. Which is actually OK, as the abi for u128 goes, but still feels less than obvious.

matklad commented on src/lsm/forest_table_iterator.zig:
```diff
@@ -157,12 +158,14 @@ fn TreeTableIteratorType(comptime Tree: type) type {
+                                    // +1 to skip past the previous table.
+                                    // (The tables are ordered by (key_max,snapshot_min).)
+                                    .snapshot_min = position.previous.snapshot_min + 1,
+                                },
```
I guess flip the order here as well? I am torn on this, but I think bitcasting data in the wrong order is more confusing than reversed lexicographic comparison

matklad commented on src/lsm/forest_table_iterator.zig:
```diff
@@ -178,7 +181,7 @@ fn TreeTableIteratorType(comptime Tree: type) type {
                 const table_order =
-                    Tree.Table.compare_keys(position.previous.key_max, table.key_max);
+                    std.math.order(position.previous.key_max, table.key_max);
                 assert(table_order.compare(.lte));
```
Maybe `switch` over `table_order` here, with an explicit unreachable branch for `ge`?

matklad commented on src/lsm/composite_key.zig:
```diff
@@ -4,16 +4,16 @@ const math = std.math;
+pub fn CompositeKeyType(comptime Field: type) type {
     assert(Field == u128 or Field == u64);
 
     return extern struct {
```
I _think_ this now wants to be a `packed` struct? Packed structs are "integres with extra semantics", and that's what we have here essentially.

batiati replied:
Packed struct simplifies the alignment, and `packet struct(IntType)` communicates the intent nicely.

Also:
- Added asserts for `padding == pad`
- Added asserts for `timestamp & tombstone_bit == 0`
- Padding now can be `void` instead of `[0]u8`

matklad reviewed (commented):
Awesome!, but I am genuinely puzzled by the `assert(!(a > b))` form, as it seems to me exactly the opposite of readable. Curious what other folks think here!

matklad commented on src/lsm/composite_key.zig:
```diff
@@ -4,67 +4,73 @@ const math = std.math;
+    );
 
-    return extern struct {
+    return packed struct(Int) {
```
Didn't realize that one can manually specify the backing integer type, that's ninfty!

matklad commented on src/lsm/composite_key.zig:
```diff
@@ -4,67 +4,73 @@ const math = std.math;
+    const Int = std.meta.Int(
+        .unsigned,
+        @bitSizeOf(u64) + @bitSizeOf(Field) + @bitSizeOf(Pad),
+    );
```
As we only support u128 and u64 fields, maybe this will be more readable as an explicit switch with hard-coded `u128` and `u256` arms? Works either way for me.

matklad commented on src/lsm/composite_key.zig:
```diff
@@ -4,67 +4,73 @@ const math = std.math;
+
+        field: Field,
+
+        padding: Pad = pad,
```
:facepalm: I think I mixed things up when I said that padding first breaks alignment. I think we actually _need_ to have padding first to preserve alignment, _because_ timestamp precedes Field. 

For `Field = u128, timestamp = u64, pad = u64`, the traditional layout would be:

```
Field     // u128
timestamp // u64
Pad       // u64
```

This layout is well-aligned, but doesn't fit little-endian comparison order we need. 

With timestamp first,  we need this layout for alignment to add up:

```
Pad       // u64
timestamp // u64
Field     // u128
```


:thinking: honestly, I think it doesn't really matter either way. 

To confuse us with even more options: I realized that we don't actually need to change composite key layout here. We are doing non-trivial masking operation in `key_from_value` anyway, and that can read like:

```
return @as(Key, value.timestamp & ~tombstone_bin) | (@as(Key, value.field) << 64)
```

This then would be completely independent on the actual layout of data on disk.

I think I am slightly in favor of this latter option -- what makes me slightly uneasy about the current state in the PR, is that our "integer keys" affects the data layout on disk somewhat (as we rearrange the order of bytes), but it is _still_ not just "take these bytes as a key", as we need to mask away the tombstone. 

But we actually don't have to change the layout on disk! `fn key_from_value` allows us to cook up the integer key in whichever way we want. So we preserve both:

* the natural, aligned `field, timestamp, padding` layout on disk
* the integer keys

I don't have a _strong_ argument for why "natural" layout is actually good though, except that it is, well, non-weird :) 

Sorry for back and forth here!

batiati replied:
> return @as(Key, value.timestamp & ~tombstone_bin) | (@as(Key, value.field) << 64)

That's a great solution. We don't particularly care about keeping the data layout on disk as a little-endian integer, in fact, I think it gives more flexibility to the LSM layout.

> Sorry for back and forth here!

Not at all!!!

batiati replied:
I restored the `CompositeKey`'s layout as an `extern struct` to be able to keep the alignment.

However, I kept `KeyMaxSnapshotMin` as a `packed struct(Int)` because we don't store it.
Just reduced the code "weirdness" by not `@bitcast`ing it around.

batiati committed:
Removes the compare_keys function.
Removes unnecessary extra comparisons performed by std.math.order when we need to check only one condition, e.g. a > b instead of std.math.order(a, b) == .gt.

batiati committed:
Add CompositeKey unit tests

batiati commented:
Sorry @matklad, just added some tests for the CompositeKey logic.
